packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -497,36 +553,55 @@ class ClipPathLayer extends ContainerLayer {
   }
 }
 
-/// A composited layer that applies a transformation matrix to its children.
+/// A composited layer that applies a given transformation matrix to its
+/// children.
+///
+/// This class inherits from [OffsetLayer] to make it one of the layers that
+/// can be used at the root of a [RenderObject] hierarchy.
 class TransformLayer extends OffsetLayer {
   /// Creates a transform layer.
   ///
-  /// The [transform] property must be non-null before the compositing phase of
-  /// the pipeline.
-  TransformLayer({
-    this.transform
-  });
+  /// The [transform] and [offset] properties must be non-null before the
+  /// compositing phase of the pipeline.
+  TransformLayer({ this.transform, Offset offset: Offset.zero }) : super(offset: offset);
 
   /// The matrix to apply.
   ///
   /// The scene must be explicitly recomposited after this property is changed
   /// (as described at [Layer]).
+  ///
+  /// This transform is applied before [offset], if both are set.
+  ///
+  /// The [transform] property must be non-null before the compositing phase of
+  /// the pipeline.
   Matrix4 transform;
 
+  Matrix4 _lastEffectiveTransform;
+
   @override
   void addToScene(ui.SceneBuilder builder, Offset layerOffset) {
-    assert(offset == Offset.zero);
-    Matrix4 effectiveTransform = transform;
+    _lastEffectiveTransform = transform;
+    if (offset != Offset.zero) {
+      _lastEffectiveTransform = new Matrix4.translationValues(offset.dx, offset.dy, 0.0)
+        ..multiply(_lastEffectiveTransform);
+    }
     if (layerOffset != Offset.zero) {
-      effectiveTransform = new Matrix4.translationValues(layerOffset.dx, layerOffset.dy, 0.0)
-        ..multiply(transform);
+      _lastEffectiveTransform = new Matrix4.translationValues(layerOffset.dx, layerOffset.dy, 0.0)
+        ..multiply(_lastEffectiveTransform);
     }
```

Comment:
It would be more efficient to add the two offsets before multiplying them into the transform (if both are present).